server/mscalendar/availability.go

```diff
@@ -115,37 +118,119 @@ func (m *mscalendar) retrieveUsersToSync(userIndex store.UserIndex, syncJobSumma
+
+	in := make(chan store.User, concurrency*4)
+	out := make(chan StatusSyncJobSummary, concurrency*4)
+	var wg sync.WaitGroup
+
+	ctxTimeout, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	defer cancel()
+
+	for i := 0; i <= concurrency; i++ {
+		go func(m mscalendar, c context.Context, w *sync.WaitGroup, in chan store.User, out chan StatusSyncJobSummary) {
+			for {
+				select {
+				case <-c.Done():
+					m.Logger.Errorf("Tiemout processing users availability")
+					break
+				case user, ok := <-in:
+					if !ok {
+						// Closed channel
+						return
+					}
+
+					js := StatusSyncJobSummary{}
+					engine, err := m.FilterCopy(withActingUser(user.MattermostUserID))
+					if err != nil {
+						m.Logger.Warnf("Not able to enable active user %s from user index. err=%v", user.MattermostUserID, err)
+						w.Done()
+						continue
+					}
+
+					calendarUser := newUserFromStoredUser(&user)
+					js.CalendarEvents, err = engine.GetCalendarEvents(calendarUser, start, end, true)
+					if err != nil {
+						js.NumberOfUsersFailedStatusChanged++
+						m.Logger.With(bot.LogContext{
+							"user": user.MattermostUserID,
+							"err":  err,
+						}).Errorf("error getting calendar events")
+						w.Done()
+						continue
+					}
+
+					out <- js
+					w.Done()
+				}
 			}
+		}(*m, ctxTimeout, &wg, in, out)
+	}
 
-			calendarUser := newUserFromStoredUser(user)
-			calendarEvents, err := engine.GetCalendarEvents(calendarUser, start, end, true)
-			if err != nil {
-				syncJobSummary.NumberOfUsersFailedStatusChanged++
-				m.Logger.With(bot.LogContext{
-					"user": u.MattermostUserID,
-					"err":  err,
-				}).Errorf("error getting calendar events")
-				continue
+	numberOfLogs := 0
+	users := []*store.User{}
+	calendarViews := []*remote.ViewCalendarResponse{}
+	for _, u := range userIndex {
+		user, err := m.Store.LoadUser(u.MattermostUserID)
+		if err != nil {
+			syncJobSummary.NumberOfUsersFailedStatusChanged++
+			if numberOfLogs < logTruncateLimit {
+				m.Logger.Warnf("Not able to load user %s from user index. err=%v", u.MattermostUserID, err)
+			} else if numberOfLogs == logTruncateLimit {
+				m.Logger.Warnf(logTruncateMsg)
 			}
-			calendarViews = append(calendarViews, calendarEvents)
+			numberOfLogs++
+			continue
 		}
+
+		// If user does not have the proper features enabled, just go to the next one
+		if !(user.Settings.UpdateStatus || user.Settings.ReceiveReminders) {
+			continue
+		}
+
+		wg.Add(1)
+		in <- *user
+
+		users = append(users, user)
 	}
+
 	if len(users) == 0 {
+		close(in)
+		close(out)
 		return users, calendarViews, errNoUsersNeedToBeSynced
 	}
 
-	if !fetchIndividually {
-		var err error
-		calendarViews, err = m.GetCalendarViews(users)
-		if err != nil {
-			return users, calendarViews, errors.Wrap(err, "not able to get calendar views for connected users")
-		}
+	go func() {
+		wg.Wait()
+		close(in)
+		// Ensure information was pulled successfully
+		time.Sleep(250 * time.Millisecond)
```

Comment:
Why is 250ms chosen specifically?

Reply:
No specific reason, just trying to check what was the problem with the tests, but also leaving some room for the `out` channel to be consumed before closing it. We can probably work with a lower value.

Reply:
The code here does seam right. If `wg.Wait()` returned, all workers should have returned. Why does the code need to wait before we close `out`?